File: nexus/hooks/buffer.py
"""
Persistent buffer for crash recovery

Acts as safety net to prevent memory loss even if all hooks fail.
Continuously buffers session context for recovery after crashes.
"""

import asyncio
import json
import threading
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, Optional


class PersistentBuffer:
    """
    SAFETY LAYER: Persistent buffer for crash recovery

    Continuously buffers session context. If all other detection
    methods fail, we can recover from this buffer.

    Buffer lifecycle:
    1. Start buffering when session starts
    2. Continuously append context
    3. Periodically flush to disk
    4. Recover after crash
    5. Clear buffer after successful storage
    """

    def __init__(self, buffer_dir: Path = None):
        """
        Initialize persistent buffer

        Args:
            buffer_dir: Directory for buffer files (default: ~/.nexus/buffer)
        """
        self.buffer_dir = buffer_dir or Path.home() / ".nexus" / "buffer"
        self.buffer_dir.mkdir(parents=True, exist_ok=True)

        self._buffers: Dict[str, dict] = {}
        self._locks: Dict[str, threading.Lock] = {}
        self._flush_interval = 10  # Flush every 10 seconds

    def start_buffering(self, agent_type: str):
        """
        Start buffering for an agent

        Args:
            agent_type: Type of agent to buffer
        """
        if agent_type not in self._buffers:
            self._buffers[agent_type] = {
                "started_at": datetime.now().isoformat(),
                "entries": [],
                "last_flush": datetime.now().isoformat(),
            }
            self._locks[agent_type] = threading.Lock()

            # Start periodic flush
            asyncio.create_task(self._periodic_flush(agent_type))

    def buffer_context(self, agent_type: str, context: Any, context_type: str = "general"):
        """
        Buffer context entry

        Args:
            agent_type: Type of agent
            context: Context to buffer (any JSON-serializable type)
            context_type: Type of context (decision, conversation, etc.)
        """
        if agent_type not in self._buffers:
            self.start_buffering(agent_type)

        with self._locks[agent_type]:
            entry = {
                "timestamp": datetime.now().isoformat(),
                "type": context_type,
                "context": context,
            }
            self._buffers[agent_type]["entries"].append(entry)

            # Auto-flush if buffer is large
            if len(self._buffers[agent_type]["entries"]) >= 10:
                self._flush_to_disk(agent_type)

    async def _periodic_flush(self, agent_type: str):
        """
        Periodically flush buffer to disk

        Runs in background
        """
        while agent_type in self._buffers:
            await asyncio.sleep(self._flush_interval)
            self._flush_to_disk(agent_type)

    def _flush_to_disk(self, agent_type: str):
        """
        Flush buffer to disk

        Thread-safe flush to disk
        """
        if agent_type not in self._buffers:
            return

        with self._locks[agent_type]:
            try:
                buffer_file = self.buffer_dir / f"{agent_type}.json"
                buffer_file.write_text(json.dumps(self._buffers[agent_type], indent=2))
                self._buffers[agent_type]["last_flush"] = datetime.now().isoformat()
            except Exception as e:
                print(f"Failed to flush buffer for {agent_type}: {e}")

    def recover_buffer(self, agent_type: str) -> Optional[dict]:
        """
        Recover buffered context after crash

        Args:
            agent_type: Type of agent to recover

        Returns:
            Buffered context or None if no buffer exists
        """
        buffer_file = self.buffer_dir / f"{agent_type}.json"

        if not buffer_file.exists():
            return None

        try:
            buffered_data = json.loads(buffer_file.read_text())
            print(f"Recovered buffer for {agent_type}: {len(buffered_data['entries'])} entries")
            return buffered_data
        except Exception as e:
            print(f"Failed to recover buffer for {agent_type}: {e}")
            return None

    def clear_buffer(self, agent_type: str):
        """
        Clear buffer after successful storage

        Args:
            agent_type: Type of agent
        """
        # Clear from memory
        if agent_type in self._buffers:
            del self._buffers[agent_type]

        if agent_type in self._locks:
            del self._locks[agent_type]

        # Clear from disk
        buffer_file = self.buffer_dir / f"{agent_type}.json"
        if buffer_file.exists():
            try:
                buffer_file.unlink()
            except Exception as e:
                print(f"Failed to clear buffer file for {agent_type}: {e}")

    def get_buffer_status(self, agent_type: str) -> Optional[dict]:
        """
        Get buffer status

        Args:
            agent_type: Type of agent

        Returns:
            Buffer status or None
        """
        if agent_type not in self._buffers:
            return None

        with self._locks[agent_type]:
            return {
                "agent_type": agent_type,
                "started_at": self._buffers[agent_type]["started_at"],
                "entries_count": len(self._buffers[agent_type]["entries"]),
                "last_flush": self._buffers[agent_type]["last_flush"],
            }

    def list_buffers(self) -> list:
        """
        List all active buffers

        Returns:
            List of buffer statuses
        """
        return [
            self.get_buffer_status(agent_type)
            for agent_type in self._buffers
        ]


class BufferEntry:
    """Single buffer entry"""

    def __init__(self, context: Any, context_type: str = "general"):
        self.timestamp = datetime.now().isoformat()
        self.type = context_type
        self.context = context

    def to_dict(self) -> dict:
        """Convert to dictionary"""
        return {
            "timestamp": self.timestamp,
            "type": self.type,
            "context": self.context,
        }

File: crates/nexus-hooks/src/buffer.rs
//! Persistent buffer for crash recovery
//!
//! The persistent buffer acts as a safety net to prevent memory loss
//! even if all hooks fail. It continuously buffers session context
//! for recovery after crashes.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
use tokio::fs;
use tokio::io::AsyncWriteExt;
use tokio::sync::RwLock;

use crate::error::{HookError, Result};
use crate::session::SessionContext;

/// Default buffer directory
pub fn default_buffer_dir() -> PathBuf {
    dirs::data_local_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("nexus")
        .join("buffer")
}

/// Buffer entry stored in memory and on disk
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BufferEntry {
    /// Timestamp when entry was created
    pub timestamp: DateTime<Utc>,

    /// Type of context
    pub context_type: String,

    /// The context data
    pub context: SessionContext,
}

/// Buffer data structure stored on disk
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BufferData {
    /// When buffering started
    pub started_at: DateTime<Utc>,

    /// Buffer entries
    pub entries: Vec<BufferEntry>,

    /// When last flushed to disk
    pub last_flush: Option<DateTime<Utc>>,

    /// Agent type
    pub agent_type: String,
}

impl BufferData {
    pub fn new(agent_type: impl Into<String>) -> Self {
        Self {
            started_at: Utc::now(),
            entries: Vec::new(),
            last_flush: None,
            agent_type: agent_type.into(),
        }
    }
}

/// Persistent buffer for crash recovery
///
/// This buffer continuously stores session context entries. If all other
/// detection methods fail, we can recover from this buffer.
///
/// # Buffer Lifecycle
///
/// 1. Start buffering when session starts
/// 2. Continuously append context entries
/// 3. Periodically flush to disk
/// 4. Recover after crash
/// 5. Clear buffer after successful storage
///
/// # Example
///
/// ```rust,no_run
/// use nexus_hooks::buffer::PersistentBuffer;
/// use nexus_hooks::session::SessionContext;
///
/// #[tokio::main]
/// async fn main() -> Result<(), Box<dyn std::error::Error>> {
///     let mut buffer = PersistentBuffer::new(None)?;
///
///     // Start buffering for an agent
///     buffer.start_buffering("claude-code").await?;
///
///     // Buffer context periodically
///     let ctx = SessionContext::new("claude-code");
///     buffer.buffer_context("claude-code", ctx, "checkpoint").await?;
///
///     // Recover after crash
///     if let Some(data) = buffer.recover_buffer("claude-code").await? {
///         println!("Recovered {} entries", data.entries.len());
///     }
///
///     // Clear after successful storage
///     buffer.clear_buffer("claude-code").await?;
///
///     Ok(())
/// }
/// ```
pub struct PersistentBuffer {
    /// Buffer directory
    buffer_dir: PathBuf,

    /// In-memory buffers by agent type
    buffers: Arc<RwLock<HashMap<String, BufferData>>>,

    /// Flush interval in seconds
    flush_interval_secs: u64,

    /// Maximum entries before auto-flush
    max_entries: usize,
}

impl PersistentBuffer {
    /// Create a new persistent buffer
    ///
    /// # Arguments
    ///
    /// * `buffer_dir` - Directory for buffer files (default: ~/.local/share/nexus/buffer)
    pub fn new(buffer_dir: Option<PathBuf>) -> Result<Self> {
        let buffer_dir = buffer_dir.unwrap_or_else(default_buffer_dir);

        // Create directory if it doesn't exist
        std::fs::create_dir_all(&buffer_dir)
            .map_err(|e| HookError::BufferError(format!("Failed to create buffer dir: {}", e)))?;

        Ok(Self {
            buffer_dir,
            buffers: Arc::new(RwLock::new(HashMap::new())),
            flush_interval_secs: 10,
            max_entries: 10,
        })
    }

    /// Set flush interval
    pub fn with_flush_interval(mut self, secs: u64) -> Self {
        self.flush_interval_secs = secs;
        self
    }

    /// Set max entries before auto-flush
    pub fn with_max_entries(mut self, max: usize) -> Self {
        self.max_entries = max;
        self
    }

    /// Start buffering for an agent
    pub async fn start_buffering(&self, agent_type: &str) -> Result<()> {
        let mut buffers = self.buffers.write().await;

        if !buffers.contains_key(agent_type) {
            buffers.insert(agent_type.to_string(), BufferData::new(agent_type));
        }

        Ok(())
    }

    /// Buffer a context entry
    pub async fn buffer_context(
        &self,
        agent_type: &str,
        context: SessionContext,
        context_type: &str,
    ) -> Result<()> {
        // Ensure buffering is started
        {
            let mut buffers = self.buffers.write().await;
            if !buffers.contains_key(agent_type) {
                buffers.insert(agent_type.to_string(), BufferData::new(agent_type));
            }
        }

        let entry = BufferEntry {
            timestamp: Utc::now(),
            context_type: context_type.to_string(),
            context,
        };

        // Add to memory buffer
        let should_flush = {
            let mut buffers = self.buffers.write().await;
            if let Some(buffer) = buffers.get_mut(agent_type) {
                buffer.entries.push(entry);
                buffer.entries.len() >= self.max_entries
            } else {
                false
            }
        };

        // Auto-flush if buffer is large
        if should_flush {
            self.flush_to_disk(agent_type).await?;
        }

        Ok(())
    }

    /// Flush buffer to disk
    pub async fn flush_to_disk(&self, agent_type: &str) -> Result<()> {
        let buffers = self.buffers.read().await;

        if let Some(buffer) = buffers.get(agent_type) {
            let buffer_file = self.buffer_dir.join(format!("{}.json", agent_type));
            let json = serde_json::to_string_pretty(buffer)
                .map_err(|e| HookError::BufferError(format!("Failed to serialize: {}", e)))?;

            let mut file = fs::File::create(&buffer_file)
                .await
                .map_err(|e| HookError::BufferError(format!("Failed to create file: {}", e)))?;

            file.write_all(json.as_bytes())
                .await
                .map_err(|e| HookError::BufferError(format!("Failed to write: {}", e)))?;

            // Update last_flush time
            drop(buffers);
            let mut buffers = self.buffers.write().await;
            if let Some(buffer) = buffers.get_mut(agent_type) {
                buffer.last_flush = Some(Utc::now());
            }
        }

        Ok(())
    }

    /// Flush all buffers to disk
    pub async fn flush_all(&self) -> Result<()> {
        let buffers = self.buffers.read().await;
        let agent_types: Vec<String> = buffers.keys().cloned().collect();
        drop(buffers);

        for agent_type in agent_types {
            self.flush_to_disk(&agent_type).await?;
        }

        Ok(())
    }

    /// Recover buffered context after crash
    pub async fn recover_buffer(&self, agent_type: &str) -> Result<Option<BufferData>> {
        let buffer_file = self.buffer_dir.join(format!("{}.json", agent_type));

        if !buffer_file.exists() {
            return Ok(None);
        }

        let content = fs::read_to_string(&buffer_file)
            .await
            .map_err(|e| HookError::BufferError(format!("Failed to read buffer: {}", e)))?;

        let data: BufferData = serde_json::from_str(&content)
            .map_err(|e| HookError::BufferError(format!("Failed to parse buffer: {}", e)))?;

        tracing::info!(
            "Recovered buffer for {}: {} entries",
            agent_type,
            data.entries.len()
        );

        Ok(Some(data))
    }

    /// Clear buffer after successful storage
    pub async fn clear_buffer(&self, agent_type: &str) -> Result<()> {
        // Clear from memory
        {
            let mut buffers = self.buffers.write().await;
            buffers.remove(agent_type);
        }

        // Clear from disk
        let buffer_file = self.buffer_dir.join(format!("{}.json", agent_type));
        if buffer_file.exists() {
            fs::remove_file(&buffer_file)
                .await
                .map_err(|e| HookError::BufferError(format!("Failed to remove buffer: {}", e)))?;
        }

        Ok(())
    }

    /// Get buffer status
    pub async fn get_buffer_status(&self, agent_type: &str) -> Option<BufferStatus> {
        let buffers = self.buffers.read().await;

        buffers.get(agent_type).map(|buffer| BufferStatus {
            agent_type: agent_type.to_string(),
            started_at: buffer.started_at,
            entries_count: buffer.entries.len(),
            last_flush: buffer.last_flush,
        })
    }

    /// List all active buffers
    pub async fn list_buffers(&self) -> Vec<BufferStatus> {
        let buffers = self.buffers.read().await;

        buffers
            .iter()
            .map(|(agent_type, buffer)| BufferStatus {
                agent_type: agent_type.clone(),
                started_at: buffer.started_at,
                entries_count: buffer.entries.len(),
                last_flush: buffer.last_flush,
            })
            .collect()
    }

    /// Check if buffer exists for agent
    pub async fn has_buffer(&self, agent_type: &str) -> bool {
        let buffers = self.buffers.read().await;
        buffers.contains_key(agent_type)
            || self
                .buffer_dir
                .join(format!("{}.json", agent_type))
                .exists()
    }
}

/// Buffer status information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BufferStatus {
    pub agent_type: String,
    pub started_at: DateTime<Utc>,
    pub entries_count: usize,
    pub last_flush: Option<DateTime<Utc>>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[tokio::test]
    async fn test_buffer_context() {
        let dir = tempdir().unwrap();
        let buffer = PersistentBuffer::new(Some(dir.path().to_path_buf())).unwrap();

        buffer.start_buffering("test-agent").await.unwrap();

        let ctx = SessionContext::new("test-agent");
        buffer
            .buffer_context("test-agent", ctx, "checkpoint")
            .await
            .unwrap();

        let status = buffer.get_buffer_status("test-agent").await.unwrap();
        assert_eq!(status.entries_count, 1);
    }

    #[tokio::test]
    async fn test_flush_and_recover() {
        let dir = tempdir().unwrap();
        let buffer = PersistentBuffer::new(Some(dir.path().to_path_buf()))
            .unwrap()
            .with_max_entries(1);

        let ctx = SessionContext::new("test-agent");

        // This should auto-flush due to max_entries = 1
        buffer.start_buffering("test-agent").await.unwrap();
        buffer
            .buffer_context("test-agent", ctx.clone(), "test")
            .await
            .unwrap();

        // Recover
        let recovered = buffer.recover_buffer("test-agent").await.unwrap();
        assert!(recovered.is_some());

        let data = recovered.unwrap();
        assert_eq!(data.entries.len(), 1);
    }

    #[tokio::test]
    async fn test_clear_buffer() {
        let dir = tempdir().unwrap();
        let buffer = PersistentBuffer::new(Some(dir.path().to_path_buf())).unwrap();

        buffer.start_buffering("test-agent").await.unwrap();

        let ctx = SessionContext::new("test-agent");
        buffer
            .buffer_context("test-agent", ctx, "test")
            .await
            .unwrap();

        buffer.flush_to_disk("test-agent").await.unwrap();
        buffer.clear_buffer("test-agent").await.unwrap();

        let status = buffer.get_buffer_status("test-agent").await;
        assert!(status.is_none());

        let recovered = buffer.recover_buffer("test-agent").await.unwrap();
        assert!(recovered.is_none());
    }

    #[tokio::test]
    async fn test_list_buffers() {
        let dir = tempdir().unwrap();
        let buffer = PersistentBuffer::new(Some(dir.path().to_path_buf())).unwrap();

        buffer.start_buffering("agent1").await.unwrap();
        buffer.start_buffering("agent2").await.unwrap();

        let buffers = buffer.list_buffers().await;
        assert_eq!(buffers.len(), 2);
    }
}
